### PR TITLE
Refactor outport driver

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- Outport handler for each driver would only work with json marshaller structures
- Messages used in each driver interface were using interfaces instead of structs

## Proposed changes
- Refactored `Driver` and `OutportHandler` to work with proto messages instead of interfaces.
- Each driver can now configure its own marshaller type(json/proto) and outport handler will fill `HeaderBytes` inside `BlockData` for `SaveBlock` & `RevertIndexedBlock`
## Testing procedure
-

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: outport/process/outportDataProvider.go [hunk 1]
```diff
@@ -150,41 +180,236 @@ func (odp *outportDataProvider) getSignersIndexes(header data.HeaderHandler) ([]
...
+
+	body, castOk := bodyHandler.(*block.Body)
+	if !castOk {
+		return nil, errCannotCastBlockBody
+	}
+
+	return body, nil
+}
+
+func (odp *outportDataProvider) getHeaderBytes(headerHandler data.HeaderHandler) ([]byte, core.HeaderType, error) {
```
Comment (mariusmihaic) [inline outport/process/outportDataProvider.go:196]: TODO: 
Move this in mx-chain-core

File: factory/status/statusComponents.go [hunk 2]
```diff
@@ -175,13 +176,26 @@ func (pc *statusComponents) epochStartEventHandler() epochStart.ActionHandler {
...
+			ShardValidatorsPubKeys: convertPubKeys(validatorsPubKeys),
+			Epoch:                  currentEpoch,
+		})
 
 	}, func(_ nodeData.HeaderHandler) {}, common.IndexerOrder)
 
 	return subscribeHandler
 }
 
+func convertPubKeys(validatorsPubKeys map[uint32][][]byte) map[uint32]*outportCore.PubKeys {
```
Comment (mariusmihaic) [inline factory/status/statusComponents.go:189]: TODO: Move this in core

File: process/block/metablock.go [hunk 3]
```diff
@@ -617,15 +617,22 @@ func (mp *metaProcessor) indexBlock(
...
+		HighestFinalBlockHash:  mp.forkDetector.GetHighestFinalBlockHash(),
 	})
 	if err != nil {
 		log.Warn("metaProcessor.indexBlock cannot prepare argSaveBlock", "error", err.Error())
 		return
 	}
-	mp.outportHandler.SaveBlock(argSaveBlock)
+	err = mp.outportHandler.SaveBlock(argSaveBlock)
+	if err != nil {
+		log.Warn("metaProcessor.outportHandler.SaveBlock cannot save block", "error", err)
```
Comment (miiu96) [inline process/block/metablock.go:632]: can you log also  the block hash and nonce

File: outport/process/outportDataProvider.go [hunk 4]
```diff
@@ -150,41 +175,201 @@ func (odp *outportDataProvider) getSignersIndexes(header data.HeaderHandler) ([]
...
+	for scrHash, txHandler := range scrs {
+		scr, castOk := txHandler.(*smartContractResult.SmartContractResult)
+		scrHashHex := getHexEncodedHash(scrHash)
+		if !castOk {
+			return nil, fmt.Errorf("%w, hash: %s", errCannotCastSCR, scrHashHex)
+		}
+
+		ret[scrHashHex] = &outportcore.SCRInfo{
+			SmartContractResult: scr,
+			FeeInfo:             newFeeInfo(),
```
Comment (ssd04) [inline outport/process/outportDataProvider.go:300]: question: FeeInfo will be set later on with the correct values, right?

File: outport/process/outportDataProvider.go [hunk 5]
```diff
@@ -150,41 +175,201 @@ func (odp *outportDataProvider) getSignersIndexes(header data.HeaderHandler) ([]
...
+		InitialPaidFee: big.NewInt(0),
 	}
 }
 
-func WrapTxsMap(txs map[string]data.TransactionHandler) map[string]data.TransactionHandlerWithGasUsedAndFee {
-	newMap := make(map[string]data.TransactionHandlerWithGasUsedAndFee, len(txs))
-	for txHash, tx := range txs {
-		newMap[txHash] = outportcore.NewTransactionHandlerWithGasAndFee(tx, 0, big.NewInt(0))
+func getScrs(scrs map[string]data.TransactionHandler) (map[string]*outportcore.SCRInfo, error) {
+	ret := make(map[string]*outportcore.SCRInfo, len(scrs))
```
Comment (ssd04) [inline outport/process/outportDataProvider.go:289]: optional, i would rename to retScrs or something, same for all functions

File: process/block/baseProcess.go [hunk 6]
```diff
@@ -1378,7 +1379,7 @@ func getLastSelfNotarizedHeaderByItself(chainHandler data.ChainHandler) (data.He
 func (bp *baseProcessor) setFinalizedHeaderHashInIndexer(hdrHash []byte) {
 	log.Debug("baseProcessor.setFinalizedBlockInIndexer", "finalized header hash", hdrHash)
```
Comment (ssd04) [inline process/block/baseProcess.go:1380]:
```suggestion
	log.Debug("baseProcessor.setFinalizedHeaderHashInIndexer", "finalized header hash", hdrHash)
```
?

File: outport/process/outportDataProvider.go [hunk 7]
```diff
@@ -150,41 +175,201 @@ func (odp *outportDataProvider) getSignersIndexes(header data.HeaderHandler) ([]
...
+}
+
+func getTxs(txs map[string]data.TransactionHandler) (map[string]*outportcore.TxInfo, error) {
+	ret := make(map[string]*outportcore.TxInfo, len(txs))
+
+	for txHash, txHandler := range txs {
+		tx, castOk := txHandler.(*transaction.Transaction)
+		txHashHex := getHexEncodedHash(txHash)
+		if !castOk {
+			return nil, fmt.Errorf("%w, hash: %s", errCannotCastTransaction, txHashHex)
```
Comment (ssd04) [inline outport/process/outportDataProvider.go:263]: usually we used error naming like `WrongTypeAssertion` for these cases, not with cast, maybe use common error from errors package with custom transaction mesage, same for the others; not very strong opinion though

File: outport/outport.go [hunk 8]
```diff
@@ -44,13 +43,51 @@ func NewOutport(retrialInterval time.Duration) (*outport, error) {
...
+
+func prepareBlockData(
+	headerBodyData *outportcore.HeaderDataWithBody,
+	driver Driver,
+) (*outportcore.BlockData, error) {
+	if headerBodyData == nil {
+		return nil, fmt.Errorf("outport.prepareBlockData error: %w", errNilHeaderAndBodyArgs)
 	}
+
+	marshaller := driver.GetMarshaller()
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply outport/outport.go:75]: in `outportcore.GetHeaderBytesAndType(marshaller, headerBodyData.Header)` there is already a check for nil marshaller

File: outport/process/outportDataProvider.go [hunk 9]
```diff
@@ -150,41 +175,201 @@ func (odp *outportDataProvider) getSignersIndexes(header data.HeaderHandler) ([]
...
+		Transactions:         txs,
+		SmartContractResults: scrs,
+		Rewards:              rewards,
+		Logs:                 logs,
+	}, nil
+}
+
+func getTxs(txs map[string]data.TransactionHandler) (map[string]*outportcore.TxInfo, error) {
+	ret := make(map[string]*outportcore.TxInfo, len(txs))
+
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply outport/process/outportDataProvider.go:258]: I prefer it this way, it's more readable and cleaner

File: process/block/shardblock.go [hunk 10]
```diff
@@ -596,16 +596,23 @@ func (sp *shardProcessor) indexBlockIfNeeded(
...
+		HighestFinalBlockHash:  sp.forkDetector.GetHighestFinalBlockHash(),
 	})
 	if err != nil {
 		log.Warn("shardProcessor.indexBlockIfNeeded cannot prepare argSaveBlock", "error", err.Error())
 		return
 	}
-	sp.outportHandler.SaveBlock(argSaveBlock)
+	err = sp.outportHandler.SaveBlock(argSaveBlock)
+	if err != nil {
+		log.Warn("shardProcessor.outportHandler.SaveBlock cannot save block", "error", err)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/shardblock.go:612]: Right, added them

File: factory/processing/processComponents.go [hunk 11]
```diff
@@ -1122,27 +1129,35 @@ func (pcf *processComponentsFactory) indexGenesisBlocks(
...
+				Header:     genesisBlockHeader,
+				HeaderHash: genesisBlockHash,
 			},
-			TransactionsPool: txsPoolPerShard[currentShardId],
-			AlteredAccounts:  alteredAccounts,
 		}
-		pcf.statusComponents.OutportHandler().SaveBlock(arg)
+		errOutport := pcf.statusComponents.OutportHandler().SaveBlock(arg)
+		if errOutport != nil {
+			log.Warn("indexGenesisBlocks.outportHandler.SaveBlock cannot save block", "error", err)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply factory/processing/processComponents.go:1164]: Right, good catch

File: outport/mock/driverStub.go [hunk 12]
```diff
@@ -45,41 +46,45 @@ func (d *DriverStub) SaveRoundsInfo(roundsInfos []*outportcore.RoundInfo) error
...
+func (d *DriverStub) FinalizedBlock(finalizedBlock *outportcore.FinalizedBlock) error {
 	if d.FinalizedBlockCalled != nil {
-		return d.FinalizedBlockCalled(headerHash)
+		return d.FinalizedBlockCalled(finalizedBlock)
 	}
 
 	return nil
 }
 
+func (d *DriverStub) GetMarshaller() marshal.Marshalizer {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply outport/mock/driverStub.go:84]: Right, added it

File: outport/notifier/eventNotifier.go [hunk 13]
```diff
@@ -60,10 +53,8 @@ func NewEventNotifier(args ArgsEventNotifier) (*eventNotifier, error) {
 	}
 
 	return &eventNotifier{
-		httpClient:      args.HttpClient,
-		marshalizer:     args.Marshaller,
-		hasher:          args.Hasher,
-		pubKeyConverter: args.PubKeyConverter,
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply outport/notifier/eventNotifier.go:66]: :pray:

File: outport/process/outportDataProvider.go (hunk 4 above)
Comment (mariusmihaic) [reply outport/process/outportDataProvider.go:300]: Yes, confirmed it with @miiu96 . It is set in `transactionsFeeProcessor`

File: outport/process/outportDataProvider.go (hunk 5 above)
Comment (mariusmihaic) [reply outport/process/outportDataProvider.go:289]: Would leave them as they are, keeping it simple

File: outport/process/outportDataProvider.go (hunk 7 above)
Comment (mariusmihaic) [reply outport/process/outportDataProvider.go:263]: There is a mixed practice here, we also use errors like `errCannotCastToBlockBody`/`ErrCannotCastUserAccountHandlerToVmCommonUserAccountHandler`

Not a strong opinion either, would leave it as it is

File: outport/process/outportDataProvider.go [hunk 14]
```diff
@@ -150,41 +175,201 @@ func (odp *outportDataProvider) getSignersIndexes(header data.HeaderHandler) ([]
...
+			return nil, fmt.Errorf("%w, hash: %s", errCannotCastReceipt, hashHex)
+		}
+
+		ret[hashHex] = tx
+	}
+
+	return ret, nil
+}
+
+func getLogs(logs []*data.LogData) (map[string]*transaction.Log, error) {
```
Comment (ssd04) [inline outport/process/outportDataProvider.go:341]: we change now from somehow ordered logs events to unordered, i'm thinking if this has an impact on the integrators, also for @miiu96

**Review by ssd04 — DISMISSED** [review]

**Review by miiu96 — APPROVED** [review]

File: outport/process/outportDataProvider.go (hunk 14 above)
Comment (miiu96) [reply outport/process/outportDataProvider.go]: yes, logs have to be sorted.

**Review by ssd04 — APPROVED** [review]

File: outport/process/outportDataProvider.go (hunk 14 above)
Comment (mariusmihaic) [reply outport/process/outportDataProvider.go]: Very good catch, fixed it

**Pull request merged**